File: nowcoder/python/get_min_max.py
#!/bin/bash python
"""
无序数组，同时查找最大值和最小值
"""

import time

class Solution(object):
	def getMinMax1(self, num):
		"""
		直接法，时间复杂度为O(n)，比较次数为2n
		"""
		if num == []:
			return 0, 0
		mmin = mmax = num[0]
		for i in num:
			if i < mmin:
				mmin = i
			elif i > mmax:
				mmax = i
		return mmin, mmax
	
	def getMinMax2(self, num):
		"""
		每次取两个数，先自行比较，然后较小值和上一次min比较，较大值和上一个max比较
		时间复杂度为O(n/2)，比较次数为n/2 * 3 = 1.5n
		"""
		if num == []:
			return 0, 0
		mmin = mmax = num[0]
		for i in range(len(num) // 2):
			curr_min = min(num[2*i], num[2*i+1])
			curr_max = max(num[2*i], num[2*i+1])
			mmin = min(mmin, curr_min)
			mmax = max(mmax, curr_max)
		# 处理长度为奇数的情况
		if len(num) % 2 == 1:
			mmin = min(mmin, num[-1])
			mmax = max(mmax, num[-1])
		return mmin, mmax
	
	def getMinMax3(self, num):
		"""
		递归法，类似归并排序，时间复杂度O(log2n)，比较次数是log级别的
		其实对应的物理意义就是两两分组pk，得出min和max
		然后依次进行...，32强，16强，...，4强，2强的淘汰赛
		"""
		def getMinMaxWithIndex(num, l, h):
			if l == h:
				return num[l], num[h]
			m = l + (h - l) // 2
			lmin, lmax = getMinMaxWithIndex(num, l, m)
			hmin, hmax = getMinMaxWithIndex(num, m + 1 ,h)
			return min(lmin, hmin), max(lmax, hmax)
		if num == []:
			return 0, 0
		return getMinMaxWithIndex(num, 0, len(num) - 1)

	def test(self, num):
		"""
		测试函数
		"""
		func_vec = [self.getMinMax1,
		            self.getMinMax2,
					self.getMinMax3]
		for func in func_vec:
			start = time.time()
			result = func(num)
			end = time.time()
			print("result: {}, times(us): {:>.5f}".format(result, (end - start)*10**6))

if __name__ == "__main__":
	num = [3, 0, 5, 8, 3, 3, 4, 9, 1]
	print(num)
	s = Solution()
	s.test(num)

File: nowcoder/cpp/get_min_max.cpp
#include <stdio.h>
#include <vector>

#define MIN(x, y) ((x) < (y) ? (x) : (y))
#define MAX(x, y) ((x) > (y) ? (x) : (y))

class Solution {
public:
	// 直接法
	void getMinMax1(std::vector<int>& num, int* min, int* max) {
		if (num.empty()) {
			*min = 0;
			*max = 0;
		}
		*min = num[0];
		*max = num[0];
		for (auto & i : num) {
			if (i < *min) {
				*min = i;
			} 
			if (i > *max) {
				*max = i;
			}
		}
	}

	// 分组比较法，每次取两个数
	void getMinMax2(std::vector<int>& num, int* min, int* max) {
		if (num.empty()) {
			*min = 0;
			*max = 0;
		}
		for (int i = 0; i < num.size() / 2; i++) {
			int curr_min = MIN(num[2*i], num[2*i+1]);
			int curr_max = MAX(num[2*i], num[2*i+1]);
			*min = MIN(*min, curr_min);
			*max = MAX(*max, curr_max);
		}
		// 处理长度为奇数的特殊情况
		if (num.size() % 2 == 1) {
			*min = MIN(*min, num.back());
			*max = MAX(*max, num.back());
		}
	}

	// 递归法，类似归并排序
	void getMinMax3(std::vector<int>& num, int* min, int* max) {
		if (num.empty()) {
			*min = 0;
			*max = 0;
		}
		getMinMaxWithIndex(num, 0, num.size() - 1, min, max);
	}

	// 归并比较的工具函数
	void getMinMaxWithIndex(std::vector<int>& num, int l, int h, int* min, int* max) {
		if (l == h) {
			*min = num[l];
			*max = num[h];
			return;
		}
		int m = l + (h - l) / 2;
		int lmin = 0, lmax = 0, hmin = 0, hmax = 0;
		getMinMaxWithIndex(num, l, m, &lmin, &lmax);
		getMinMaxWithIndex(num, m + 1, h, &hmin, &hmax);
		*min = MIN(lmin, hmin);
		*max = MAX(lmax, hmax);
	}
	
    // 测试函数
	void test(std::vector<int>& num) {
		for (auto & func : this->func_vec_) {
			int min = 0, max = 0;
			(this->*func)(num, &min, &max);
			printf("result: (%d, %d)\n", min, max);
		}
	}

private:
    typedef void (Solution::*func_ptr)(std::vector<int>&, int*, int*);
	std::vector<func_ptr> func_vec_ = {&Solution::getMinMax1,
	                                   &Solution::getMinMax2,
									   &Solution::getMinMax3};
};

int main(int argc, char* argv[])
{
	std::vector<int> num = {3, 0, 5, 8, 3, 3, 4, 9, 1};
	Solution s;
	s.test(num);
	return 0;
}